src/bvh/bvh.rs

```diff
@@ -0,0 +1,444 @@
+    fn pretty_print(&self, nodes: &Vec<BVHNode>, depth: usize) {
+        let padding: String = repeat(" ").take(depth).collect();
+        match *self {
+            BVHNode::Node { child_l, child_r, .. } => {
+                println!("{}child_l", padding);
+                nodes[child_l].pretty_print(nodes, depth + 1);
+                println!("{}child_r", padding);
+                nodes[child_r].pretty_print(nodes, depth + 1);
+            }
+            BVHNode::Leaf { shape, .. } => {
+                println!("{}shape\t{:?}", padding, shape);
+            }
+            BVHNode::Dummy => {
+                println!("{}dummy node!", padding);
+            }
+        }
+    }
+
+    /// Traverses the [`BVH`] recursively and insterts shapes which are hit with a
+    /// high probability by `ray` into the [`Vec`] `indices`.
+    ///
+    /// [`BVH`]: struct.BVH.html
+    /// [`Vec`]: https://doc.rust-lang.org/std/vec/struct.Vec.html
+    ///
+    pub fn traverse_recursive(&self, nodes: &Vec<BVHNode>, ray: &Ray, indices: &mut Vec<usize>) {
+        match *self {
+            BVHNode::Node { ref child_l_aabb, child_l, ref child_r_aabb, child_r, .. } => {
+                if ray.intersects_aabb(child_l_aabb) {
+                    nodes[child_l].traverse_recursive(nodes, ray, indices);
+                }
+                if ray.intersects_aabb(child_r_aabb) {
+                    nodes[child_r].traverse_recursive(nodes, ray, indices);
+                }
+            }
+            BVHNode::Leaf { shape, .. } => {
+                indices.push(shape);
+            }
+            BVHNode::Dummy => {
+                panic!("Dummy node found during BVH traversal!");
+            }
+        }
+    }
+}
+
+/// The [`BVH`] data structure. Only contains the root node of the [`BVH`] tree.
+///
+/// [`BVH`]: struct.BVH.html
+///
+pub struct BVH {
+    /// The list of nodes of the [`BVH`].
+    ///
+    /// [`BVH`]: struct.BVH.html
+    ///
+    pub nodes: Vec<BVHNode>,
+}
+
+impl BVH {
+    /// Creates a new [`BVH`] from the `shapes` slice.
+    ///
+    /// [`BVH`]: struct.BVH.html
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use bvh::aabb::{AABB, Bounded};
+    /// use bvh::bvh::{BVH, BHShape};
+    /// use
+    /// use bvh::nalgebra::{Point3, Vector3};
+    /// # impl BHShape for AABB {
+    /// #     fn set_bh_node_index(&mut self, index: usize) { }
+    /// #     fn bh_node_index(&self) -> usize { 0 }
+    /// # }
+    /// #
+    /// # fn create_bounded_shapes() -> Vec<AABB> {
+    /// #     let mut shapes = Vec::new();
+    /// #     let offset = Vector3::new(1.0, 1.0, 1.0);
+    /// #     for i in 0..1000u32 {
+    /// #         let position = Point3::new(i as f32, i as f32, i as f32);
+    /// #         shapes.push(AABB::with_bounds(position - offset, position + offset));
+    /// #     }
+    /// #     shapes
+    /// # }
+    ///
+    /// let mut shapes = create_bounded_shapes();
+    /// let bvh = BVH::build(&mut shapes);
```

Comment:
The doctests are rather broken at the moment and subject to a change before !14 gets ready to be merged.